stdlib/fun.mli

```diff
@@ -31,3 +31,25 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+val protect : finally:(unit -> unit) -> (unit -> 'a) -> 'a
+(** [protect ~finally work] invokes [work ()] and then [finally ()]
+    before [work ()] returns with its value or an exception. In the
+    latter case the exception is re-raised after [finally ()]. If
+    [finally ()] raises an exception, then the exception
+    {!Finally_raised} is raised instead.
+
+    [protect] can be used to enforce local invariants whether the
+    function returns normally or raises an exception. However, it does
+    not protect against unexpected exceptions raised inside [finally
+    ()] such as Out_of_memory, Stack_overflow, or asynchronous
```

Comment:
In the doc strings here and below the exceptions you mention could be linkified by writing `{!Out_of_memory}, {!Stack_overflow}, {!Sys.Break}`.